src/ecosystem/corals/growth.jl

```diff
@@ -669,47 +592,48 @@ function settler_DHW_tolerance!(
     groups, sizes, _ = axes(c_mean_t_1)
 
     sink_loc_ids::Vector{Int64} = findall(k_area .> 0.0)
-
     source_locs::BitVector = falses(length(k_area))  # cache for source locations
-    reproductive_sc::BitVector = falses(length(sizes))  # cache for reproductive size classes
 
-    for sink_loc in sink_loc_ids
-        if sum(@views(settlers[:, sink_loc])) .== 0.0
+    # Number of reproductive size classes for each group
+    n_reproductive = sum(view(fec_params_per_m², :, :) .> 0.0; dims=2)
+    reproductive_sc::BitVector = falses(sizes)  # cache for reproductive size classes
+
+    @inbounds for sink_loc in sink_loc_ids
+        sink_settlers = view(settlers, :, sink_loc)'
+        if sum(sink_settlers) .== 0.0
             # Only update locations where recruitment occurred
             continue
         end
 
+        # Find sources for each sink
         # Note: source locations can include the sink location (self-seeding)
-        source_locs .= @view(tp[:, sink_loc]) .> 0.0
+        source_locs .= view(tp.data, :, sink_loc) .> 0.0
 
-        # Calculate contribution to cover to determine weights for each species/group
-        w = @views settlers[:, sink_loc]' .* tp.data[source_locs, sink_loc]
-        w_per_group = w ./ sum(w; dims=1)
-        replace!(w_per_group, NaN => 0.0)
+        # Calculate contribution to cover to determine weights for each functional group
+        w::Matrix{Float64} = sink_settlers .* view(tp.data, source_locs, sink_loc)
```

Comment:
The tricky thing is the number of sources can change.

But maybe there is a way of preallocating and then resizing. Maybe it will avoid excessive triggering of the GC... let me have a think.

Reply:
Couldn't think of a way to cache the weights matrix because the number of source locations change as I said.
I did play with the idea that the number of source locations are constant, but this is only true where we use mean connectivity. If we move to variable/dynamic connectivity data, then we'd need this implementation anyway.

I did think of some other tweaks which helped a tiny bit though:

![image](https://github.com/user-attachments/assets/db1afbad-0417-47ff-b81a-92f5f549d742)